Create ADR for installing only selected CRDs (#2973)
Azure/azure-service-operator: docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md

      (such as a small `kind` cluster or a `Free` tier AKS cluster).
2. Seamless upgrade experience for existing users.
    * If the user has installed a previous version of ASO and upgrade to a new version, whatever 
      resources they had installed should continue to function. The previous default was "everything".
3. Seamless upgrade experience going forward.
    * If a user selects a subset of CRDs when initially installing ASO, they should get that exact 
      subset automatically during upgrade unless they say otherwise.
    * No "oops I forgot to select any CRDs, so none were installed"
    * No "oops I forgot to filter CRDs, so they were all installed"
    * Exactly the CRDs the user originally specified should be upgraded, and only those CRDs.
4. No CRDs uninstalled ever.
    * CRD uninstallation is dangerous. If users want to remove CRDs they can do it manually. 
    * TODO: A reduction in the list of CRDs to install should result in only the new list CRDs of CRDs getting new versions?
      Or maybe it needs to actually upgrade those versions too even though they weren't specified, to maintain a consistent
      storage version?

> ### Question 1: How to identify the subset of resources to install by default?

Reviewer: ```suggestion
### Question 1: What to install by default?
```

As mentioned above, a key goal is a reduction in the default number of CRDs installed. That means we somehow need to 
choose which CRDs we will install and which CRDs we will not.

> #### Answer 1a: We handpick a set of resources we feel are "critical" and install just those by default
> 

Reviewer: ```suggestion
#### Answer 1a: Handpicked Default

We handpick a set of resources we feel are "critical" and install just those by default

```

Author: Done

Pros:
* Has at least a chance of working for some user-scenarios out of the box.
* Ensures that resources which are extremely likely to be needed (such as Microsoft.Resources/resourceGroup) are included 
  by default.

Cons:
* Arbitrary.
* May result in higher kube-apiserver memory consumption than is strictly needed for users as it is likely to include some
  resources which they may not use.

> #### Answer 1b: We install nothing by default
> 

Reviewer: ```suggestion
#### Answer 1b: Nothing by default

We install nothing by default

```

Pros:
* Users choose exactly what they need.

Cons:
* The operator is nonfunctional out of the box, which is a poor new-user experience.

> #### Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it

Reviewer: ```suggestion
#### Answer 1c: Mandatory configuration

The selection of CRDs to install is required and the operator pod will not launch without it
```

Can also use [Helm required values](https://stackoverflow.com/questions/53866196/how-best-to-say-a-value-is-required-in-a-helm-chart)
to ensure that the value is required at the Helm chart level, although making it required may run afoul of requirement #3.

Pros:
* Users choose exactly what they need.

Cons:
* The operator is nonfunctional out of the box, but at least it's clearly nonfunctional and has a good error why.

### Question 2: How do users specify what CRDs to install?

> #### Answer 2a: using our existing --crd-pattern command line parameter on the operator pod

Reviewer: ```suggestion
#### Answer 2a: Use the existing command line parameter

Continue using the existing `--crd-pattern` command line parameter on the operator pod.
```

Pros:
* Exists today

Cons:

> * Command-line isn't preserved between upgrades, which fails requirement #3 (seamless upgrade experience).

Reviewer: ```suggestion
* Command-line isn't preserved between upgrades, which may fail requirement #3 (seamless upgrade experience).
```

> #### Answer 2b: Using a ConfigMap

Reviewer: ```suggestion
#### Answer 2b: Use a ConfigMap
```

A `ConfigMap` would persist between upgrades and so in theory could be used to remember what was installed previously
without the user needing to re-specify it.

One challenge is that in the Helm context if the `ConfigMap` is created by Helm, Helm will overwrite it in 
subsequent upgrades. This could be dealt with by using a 
[pre-install hook](https://stackoverflow.com/questions/55503893/helm-patch-default-service-account). We would create and
patch the `ConfigMap` in the pre-install hook.

> ConfigMap structure:

Reviewer: ```suggestion
One possible configMap structure:
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: managed-resources
  namespace: azureserviceoperator-system
data:
  "all": ""  # magic key meaning "*" which isn't supported in keys
  microsoft.resources: "*"

>   microsoft.storage: "StorageAccount,StorageAccountsBlobService"

Reviewer 2: We might need handling somewhere to check if parent CRD is installed or not. As for in this example, what if a new user, does something like `microsoft.storage: "StorageAccountsBlobService"` and assume it'll include StorageAccount or if its not required.

Author: I'll add an open questions section and put this there. My current thinking on this is that we avoid this complexity and strongly recommend users install entire providers such as `compute/*`.

There are other links to consider too:
- Owners (i.e. resourceGroup owns a lot of things even though it's not in their group)
- Cross resource references: i.e. compute references networking. This is pretty involved to work out automatically and there's also no guarantee that just because a resource _can_ link to a Subnet or a LoadBalancer that the customer actually has that use-case.

```

The following `patch` commands could be run either directly by the user or by our pre-install hook:

* Addition of new resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --patch '{"data":{"microsoft.storage":"*"}}'`
* Removal of old resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --type=json -p='[{"op": "remove", "path": "/data/microsoft.resources"}]'`

If nothing is specified as part of the upgrade, the set of supported resources will not change.

If the `ConfigMap` doesn't exist when we go to upgrade, we can assume we're coming from an older ASO version that didn't

> support the `ConfigMap` and default to creating one that signals to install all resources.

Reviewer: This would put us back into the high memory use scenario.

Author: Yes but it'd be a breaking change IMO to do anything other than this, right?

Reviewer: Yup. I just figure it's worth calling out as a Con for this approach.

Author: Note I'm specifically here referring to just upgrade scenarios. New installs user would need to choose still. I'll call that out as well.

Author: Added the con

  to enable ASO as an Extension or Addon, as there everything must be done through Helm. For users of Helm itself we likely
  can just tell them to run a `kubectl` command afterwards to create the `ConfigMap`. That does make the Helm install 
  experience worse though.
* We may be able to create the `ConfigMap` as empty given Helms use of
  [3 way strategic merge](https://helm.sh/docs/faq/changes_since_helm2/#improved-upgrade-strategy-3-way-strategic-merge-patches)
  and then patch it afterwards.
* Is it OK that _providers_ would be patchable, but you would need to include all the resources when patching something 
  like `microsoft.storage: "StorageAccount,StorageAccountsBlobService"`? We could just disallow individual resource selection
  for now, although that might eventually become a problem for certain large providers (`Microsoft.Network` comes to mind).

See also Helm [post rendering](https://helm.sh/docs/topics/advanced/#post-rendering), which we could also consider.
pre-install hook (or post-install) seems easier though.

#### Answer 2c: Using a CRD

We investigated doing this originally for [Helm chart size limitations](../ADR-2023-02-Helm-Chart-Size-Limitations)

Pros:

> * Slightly more structured than a `ConfigMap`.

Reviewer: ```suggestion
* Significantly more structured than a `ConfigMap`.
```

Author: Dropped the qualifier and said "More structured than a `ConfigMap`"

Cons:
* Chicken and egg problem of using the CRD that's part of the operator to configure the operator behavior. 
  Helm has a very hard time creating a CRD _and_ an instance of the CRD in the same deployment.
* May need to implement collection strategic merge behavior that already exists for `ConfigMap`.

## Decision

For question #1: [Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it](#answer-1c--selection-of-crds-to-install-is-required-and-the-operator-pod-will-not-launch-without-it)

> For question #2: [Answer 2b: Using a ConfigMap](#answer-2b--using-a-configmap)

Reviewer: With the separation of question 2 into "how to configure" and "how to ensure consistency" (as mentioned elsewhere), I think we take a different approach:

* Mandatory configuration
* Retain Command line parameter
* Always support CRDs that have already been installed

Author: Not sure exactly what "Retain command line parameter" means - does it just mean "use the option we have now" or are you talking about some cmdline parameter that is retained between deployments somehow automatically?

I agree with your argument towards "Always support CRDs that have already been installed", but I think that actually forces us away from "Mandatory configuration", at least in the sense of Helm required parameters or anything like that.

I think what we want is:
- Operator exits with a clear reason (visible on Pod status) if no CRDs are configured.
- Cmdline parameter to configure which _new_ CRDs to install
- All existing CRDs are automatically upgraded to the latest version, even if not specified in cmdline parameter.

If we have some form of mandatory configuration _and_ the "existing CRDs" behavior it's not clear what the user should put into the mandatory configuration if they want to just use the resources they were already using.

Pros:
* More structured than a `ConfigMap`.

Cons:
* Chicken and egg problem of using the CRD that's part of the operator to configure the operator behavior.
  Helm has a very hard time creating a CRD _and_ an instance of the CRD in the same deployment.
* May need to implement collection strategic merge behavior that already exists for `ConfigMap`.

### Question 3: How do we ensure consistent configuration as ASO is upgraded

#### Answer 3a: Require the user to be consistent when upgrading

The user would need re-specify what resources they want every time they upgrade

Pros:
* Easy for us

Cons:
* Very easy to make a mistake and accidentally leave a resource off they want.

#### Answer 3b: Use a configuration technique that persists across upgrades

Pros:
* Satisfies the requirement.

Cons:
* Would augment/complicate the approach we already have
* More complex for us to implement, especially with Helm. Helm owns any resource it creates, 
  and deletes it if not specified on the next upgrade. See [Using a ConfigMap](#answer-2b--use-a-configmap) for some 
  workarounds.

> 

Author: Add variant: "Store the configuration independently..."

#### Answer 3c: Query existing CRDs and always support those

Combined with some form of non-persistent storage such as [command line](#answer-2a--using-the-existing-command-line-parameter)
we would query all existing CRDs and install/upgrade them, while adding whatever new CRDs were specified in 
the non-persistent storage.

Pros:
* Automatically supports seamless upgrades.
* Installed resources are always upgrade to the latest version, even if user has forgotten about them.
* Could be built onto existing cmdline configuration.
* Hard for users to get wrong.

> 

Author: + Single source of truth

Cons:
* ???

## Decision

### Answer to [question #1: What to install by default?](#question-1--what-to-install-by-default)

A combination of  [Answer 1b: Nothing by default](#answer-1b--nothing-by-default) and 
[Answer 1c: Mandatory configuration](#answer-1c--mandatory-configuration).

Specifically, the operator pod should exit with an error if there are no CRDs configured in `--crd-pattern` _and_ no
existing CRDs in the cluster. If a `--crd-pattern` is specified, or if there are already existing ASO CRDs in the cluster
then the pod will launch without an error, install/upgrade the required CRDs and then start watches on those CRDs.

### Answer to [question #2: How do users specify what CRDs to install?](#question-2--how-do-users-specify-what-crds-to-install)

[Answer 2a: Using the existing command line parameter](#answer-2a--using-the-existing-command-line-parameter)

This parameter will come to mean: "New CRDs to install to the cluster".

### Answer to [question #3: How do we ensure consistent configuration as ASO is upgraded?](#question-3--how-do-we-ensure-consistent-configuration-as-aso-is-upgraded)

> [Answer 3c: Query existing CRDs and always support those](#answer-3c--query-existing-crds-and-always-support-those)

Author: Note: We need to deal with the users-manage-their-own CRDs case, explain what we do for that here.

Author: Confirm that in the "we dont manage CRDs case, we still have GET CRDs permissions"

Author: Updated documentation to cover this case. Fundamentally, it should work exactly as planned. See the document which outlines the basic algorithm

### Open questions and their answers

#### Should we do "graph-based" inclusions of resources?

If a user imports `microsoft.storage/storageAccountsBlobService` should we automatically determine that they 
also want `microsoft.storage/storageAccounts` since that's the `owner` type?

Conclusion: No, users must list all resources they want explicitly. We strongly recommend using full provider imports such as 
`microsoft.networking/*` rather than tightly scoping to specific resources.

## Status

> Proposed.

Author: We need to document:
1. What the CRD uninstall process is (especially for users from v2.0.0 where they have everything).
2. Explain how to actually select new CRDs
3. Explain how CRDs are managed/installed.

Author: Have made notes of all this. Documentation will be included in the implementation PR.

> Proposed.

Author: Open question, even without a predefined set of "default" resources, do we want a "default" set in the ASO install command on the readme so that people know what to do?

Consensus seems to be: yes.

Author: Made a note of this. Docs will be in the implementation PR
